File: frontend/lib/services/product_service.dart
// lib/services/product_service.dart
import 'dart:convert';

import 'package:frontend/models/product.dart';
import 'package:http/http.dart' as http;

class ProductService {
  final String baseUrl = 'http://localhost:8000/api/v1';
  final http.Client client = http.Client();

  Future<List<Product>> getProducts() async {
    final response = await client.get(Uri.parse('$baseUrl/products'));
    if (response.statusCode == 200) {
      List<dynamic> data = json.decode(response.body);
      return data.map((json) => Product.fromJson(json)).toList();
    }
    throw Exception('Failed to load products');
  }

  Future<Product> createProduct(Product product) async {
    final response = await client.post(
      Uri.parse('$baseUrl/products'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'name': product.name,
        'description': product.description,
        'price': product.price,
      }),
    );
    if (response.statusCode == 200) {
      return Product.fromJson(json.decode(response.body));
    }
    throw Exception('Failed to create product');
  }

  Future<Product> updateProduct(Product product) async {
    final response = await client.put(
      Uri.parse('$baseUrl/products/${product.productId}'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'name': product.name,
        'description': product.description,
        'price': product.price,
      }),
    );
    if (response.statusCode == 200) {
      return Product.fromJson(json.decode(response.body));
    }
    throw Exception('Failed to update product');
  }

  Future<void> deleteProduct(int productId) async {
    final response = await client.delete(
      Uri.parse('$baseUrl/products/$productId'),
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to delete product');
    }
  }
}

File: frontend/lib/services/product-service.ts
import { fetchData, postData, putData, deleteData } from "@/lib/api"
import type { ProductResponse, ProductCreate, ProductUpdate } from "@/lib/types"

const PRODUCTS_ENDPOINT = "/products"

export const getProducts = async (search?: string): Promise<ProductResponse[]> => {
  const endpoint = search ? `${PRODUCTS_ENDPOINT}?search=${encodeURIComponent(search)}` : PRODUCTS_ENDPOINT
  return fetchData<ProductResponse[]>(endpoint)
}

export const getProduct = async (productId: number): Promise<ProductResponse> => {
  return fetchData<ProductResponse>(`${PRODUCTS_ENDPOINT}/${productId}`)
}

export const createProduct = async (product: ProductCreate): Promise<ProductResponse> => {
  return postData<ProductResponse>(PRODUCTS_ENDPOINT, product)
}

export const updateProduct = async (productId: number, product: ProductUpdate): Promise<ProductResponse> => {
  return putData<ProductResponse>(`${PRODUCTS_ENDPOINT}/${productId}`, product)
}

export const deleteProduct = async (productId: number): Promise<void> => {
  return deleteData(`${PRODUCTS_ENDPOINT}/${productId}`)
}

